lib/seahorn/PredicateAbstraction.cc

```diff
@@ -0,0 +1,370 @@
+#include "seahorn/PredicateAbstraction.hh"
+#include "seahorn/HornifyModule.hh"
+#include "seahorn/HornClauseDBTransf.hh"
+#include "seahorn/HornClauseDB.hh"
+
+#include "llvm/IR/Function.h"
+#include "llvm/Support/CommandLine.h"
+
+#include "ufo/Expr.hpp"
+#include "ufo/Smt/Z3n.hpp"
+#include "ufo/Smt/EZ3.hh"
+#include <vector>
+#include <boost/logic/tribool.hpp>
+#include "seahorn/HornClauseDBWto.hh"
+#include <algorithm>
+
+#include "ufo/Stats.hh"
+
+using namespace llvm;
+
+namespace seahorn
+{
+	char PredicateAbstraction::ID = 0;
+
+	std::map<Expr, int> PredicateAbstraction::predToBiNumMap;
+    std::map<Expr, Expr> PredicateAbstraction::oldToNewPredMap;
+    std::map<Expr, Expr> PredicateAbstraction::currentCandidates;
+
+	bool PredicateAbstraction::runOnModule (Module &M)
+	{
+		HornifyModule &hm = getAnalysis<HornifyModule> ();
+
+		//load the Horn clause database
+		auto &db = hm.getHornClauseDB ();
+		db.buildIndexes ();
+
+		//build the wto
+		HornClauseDBCallGraph callgraph(db);
+		HornClauseDBWto db_wto(callgraph);
+		db_wto.buildWto();
+
+		//guess candidates
+		guessCandidate(db);
+
+		//run main algorithm
+		runOnDB(db);
+
+
+		return false;
+	}
+
+	void PredicateAbstraction::getAnalysisUsage (AnalysisUsage &AU) const
+	{
+	    AU.addRequired<HornifyModule> ();
+	    AU.setPreservesAll();
+	}
+
+	void PredicateAbstraction::runOnDB(HornClauseDB &db)
+	{
+		HornClauseDB new_DB(db.getExprFactory ());
+		int index = 0;
+
+		for(Expr rel : db.getRelations())
+		{
+			predToBiNumMap.insert(std::pair<Expr, int>(rel, index));
+			index++;
+		}
+
+		for(Expr rel : db.getRelations())
+		{
+			outs() << "OLD PRED: " << *rel << "\n";
+			ExprVector new_args;
+			//Push fdecl name
+			new_args.push_back(bind::fname(rel));
+			//Push boolean types
+			for(int i=0; i < db.getRelations().size(); i++)
+			{
+				new_args.push_back(mk<BOOL_TY>(rel->efac()));
+			}
+			Expr new_rel = mknary<FDECL>(new_args);
+			outs() << "NEW PRED: " << *new_rel << "\n";
+			new_DB.registerRelation(new_rel);
+			oldToNewPredMap.insert(std::pair<Expr, Expr>(rel, new_rel));
+		}
```

Comment:
@Chenguang-Zhu , this loop seems wrong. You should iterate over number of predicates, not number of relations in the database. 

It is useful to give the new relation a unique name so that, if desired, old and new relations can be mixed together in one database. You can use `variant::variant (bind::fname(rel), 1)` for example.